### PR TITLE
Creating and profiling CDF and Histograms

### PR DESCRIPTION
**Pull request by uwieder**

### CODE REVIEW

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/BasicColStat.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+public class BasicColStat {
+    private final int MOMENT_NUM;
+    private double min;
+    @Nullable private Object minObject;
+    private double max;
+    @Nullable private Object maxObject;
+    private final double moments[];
+    private long size;
+
+    public BasicColStat(int momentNum) {
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/BasicColStat.java]: I thought we agreed that 0 is legal

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/BasicColStat.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        if (momentNum < 1)
+            throw new IllegalArgumentException("number of moments cannot be positive");
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/BasicColStat.java]: The error message seems wrong.

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/BasicColStat.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
+package org.hiero.sketch.spreadsheet;
+
+import org.hiero.sketch.table.api.IColumn;
+import org.hiero.sketch.table.api.IMembershipSet;
+import org.hiero.sketch.table.api.IRowIterator;
+import org.hiero.sketch.table.api.IStringConverter;
+import javax.annotation.Nullable;
+
+public class BasicColStat {
+    private final int MOMENT_NUM;
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/BasicColStat.java]: In general we use all caps only for enums and constants. I would suggest camelCase: momentCount.

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/BasicColStat.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
+package org.hiero.sketch.spreadsheet;
+
+import org.hiero.sketch.table.api.IColumn;
+import org.hiero.sketch.table.api.IMembershipSet;
+import org.hiero.sketch.table.api.IRowIterator;
+import org.hiero.sketch.table.api.IStringConverter;
+import javax.annotation.Nullable;
+
+public class BasicColStat {
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/BasicColStat.java]: Please add a JavaDoc comment

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/BasicColStat.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        this.MOMENT_NUM = 2;
+        this.moments = new double[this.MOMENT_NUM];
+        this.min = Double.POSITIVE_INFINITY;
+        this.max = Double.NEGATIVE_INFINITY;
+    }
+
+    public double getMin() { return this.min; }
+
+    public Object getMinObject() {
+        assert this.minObject != null;
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/BasicColStat.java]: This assert should not be here.
Instead the function should be tagged as @Nullable.

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/BasicColStat.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+    public double getMax() { return this.max; }
+
+    public Object getMaxObject() {
+        assert this.maxObject != null;
+        return this.maxObject;
+    }
+
+    /**
+     *
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/BasicColStat.java]: extra empty line.

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/BasicColStat.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+     * @param i a number in {0,1,...,MOMENT_NUM}
+     * @return the i'th moment which is the sum of x^i
+     */
+    public double getMoment(int i) {
+        if ((i < 1) || (i > this.MOMENT_NUM))
+            throw new IllegalArgumentException("moment requested doesn't exist");
+        return this.moments[i - 1];
+    }
+
+    public long getSize() { return this.size; }
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/BasicColStat.java]: This deserves a JavaDoc comment. This is the size of what?
Perhaps it should be called something else, e.g. rowCount?

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/BasicColStat.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+    public void createStats (final IColumn column, final IMembershipSet membershipSet,
+    @Nullable final IStringConverter converter) {
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/BasicColStat.java]: indentation

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/BasicColStatSketch.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+import javax.annotation.Nullable;
+
+public class BasicColStatSketch implements ISketch<ITable, BasicColStat> {
+    final String colName;
+    @Nullable
+    final IStringConverter converter;
+    final double rate;
+    final int momentNum;
+
+    public BasicColStatSketch( String colName, @Nullable IStringConverter converter) {
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/BasicColStatSketch.java]: Extra space after (

File: hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/BasicColStatSketch.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        return result;
+    }
+
+    @Override
+    public BasicColStat zero() {
+        return new BasicColStat(momentNum);
+    }
+
+    @Override
+    public BasicColStat add(@Nullable final BasicColStat left, @Nullable final BasicColStat right) {
```
Comment (mbudiu-vmw) [inline hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/BasicColStatSketch.java]: Does this need @nullable?
If yes, then zero does too.

File: hieroplatform/src/test/java/org/hiero/sketch/CDFTest.java [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+               this.dataSet.blockingSketch(new BasicColStatSketch(this.colName, null));
+    }
+
+    Histogram1DLight prepareCDF(int width, int height, boolean useSampling) {
+        BucketsDescriptionEqSize bDec  =
+                new BucketsDescriptionEqSize(this.colStat.getMin(), this.colStat.getMax(), width);
+        double sampleSize  =  2 * height * height * width;
+        double rate = sampleSize / this.colStat.getSize();
+        if ((rate > 0.1) || (!useSampling))
+            rate = 1.0; //no use in sampling
```
Comment (mbudiu-vmw) [inline hieroplatform/src/test/java/org/hiero/sketch/CDFTest.java]: you should leave a space after //

File: hieroplatform/src/test/java/org/hiero/sketch/CDFTest.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+        return this.dataSet.blockingSketch(new Hist1DSketch(bDec, this.colName, null, rate));
+    }
+
+    @Test
+    public void HistE2E() {
+        CDFTest cdftest = new CDFTest();
+        long startTime, endTime;
+        startTime = System.nanoTime();
+        cdftest.prepareCDF(1000, 1000, false);
```
Comment (mbudiu-vmw) [inline hieroplatform/src/test/java/org/hiero/sketch/CDFTest.java]: Don't we already have a test method for timing?

**Comment by uwieder** [comment]
Made the changes

**Pull request merged**